### PR TITLE
Add quantizer to contrib

### PR DESCRIPTION
**Pull request by chrisib**
Adds a 5-output quantizer capable of either triggered or continuous operation.

### CODE REVIEW

File: software/contrib/quantizer.py
Comment (awonak) [inline software/contrib/quantizer.py:464]: [discussion] Is it necessary to include this sleep? In my experience adding a sleep may cause the script to feel sluggish and not very responsive.

**Review by awonak — COMMENTED** [review]
Nice work! I'm looking forward to trying out this script. Don't forget to add a brief description to the contrib folder's README too!

https://github.com/Allen-Synthesis/EuroPi/blob/main/software/contrib/README.md?plain=1

File: software/contrib/quantizer.py
Comment (chrisib) [reply software/contrib/quantizer.py:470]: Whoops! That's a leftover from some earlier code that apparently didn't get completely deleted.  `self.asleep` serves no purpose and should have been deleted.

File: software/contrib/quantizer.py
Comment (chrisib) [reply software/contrib/quantizer.py:464]: I didn't notice it feeling sluggish in my testing. It's there mostly to ensure that the trigger signal generated by the continuous quantization mode is on for at least 10ms.

File: software/contrib/quantizer.py
Comment (chrisib) [reply software/contrib/quantizer.py:46]: Done.

File: software/contrib/quantizer.py
Comment (chrisib) [reply software/contrib/quantizer.py]: Should be fixed now.

File: software/contrib/quantizer.py
Comment (awonak) [reply software/contrib/quantizer.py:471]: Ah, I see, thanks for clarifying. Another strategy you can use here instead of sleep is to rely on the duration of time that has passed since the trigger went high. You could add code like this in the appropriate places:

```python
# Some code that starts a trigger
cv.on()
self._trigger_deadline = time.ticks_add(time.ticks_ms(), self.TRIGGER_DURATION)

...

# If trigger deadline has passed, then turn the trigger cv off
if time.ticks_diff(time.ticks_ms(), self._trigger_deadline ) > 0:
    cv.off()
```

File: software/contrib/quantizer.py
Comment (awonak) [inline software/contrib/quantizer.py:476]: [required] Sorry I missed this the first time, MicroPython is a little funny with time math. The recommended usage is [ticks_diff](https://docs.micropython.org/en/latest/library/time.html#time.ticks_diff) because the pico internal clock could wrap around which would cause this math to fail unexpectedly.

Comment (chrisib) [reply software/contrib/quantizer.py:476]: Fair enough. This should be fixed in 631928a

File: software/contrib/quantizer.py
Comment (chrisib) [reply software/contrib/quantizer.py:471]: I'd considered that, but the problem with this approach is that we'd need to block continuous quantization until the trigger has elapsed. Otherwise the note could change twice within the duration of a single trigger's pulse width.  Blocking quantization until the previous trigger has cleared is basically the same thing as a sleep anyway, just more complicated to write.

Have you had a chance to test the script as-is with the sleep?  As I said, I didn't notice any sluggishness in my testing, but maybe I'm not as familiar with what "sluggish" feels like on the EuroPi.

File: software/contrib/quantizer.py
Comment (chrisib) [reply software/contrib/quantizer.py]: ...er, 80262fc.  It's important to make sure the file is saved before you commit...

Comment (awonak) [reply software/contrib/quantizer.py]: Thanks for the quick fix! Let's also use the `ticks_diff` and use `time_ms` to make sure we're using correct values in the logic as well as when we request the current time. I also suggest updating the const name to use the `MS` suffix to match the unit of time. (Double check my work, I'm _pretty_ sure  this is the correct order of values in ticks_diff)

```suggestion
            if time.ticks_diff(self.last_interaction_time, time.now_ms()) > SCREENSAVER_TIMEOUT_MS:
```
Where `SCREENSAVER_TIMEOUT_MS ` is:
```python
SCREENSAVER_TIMEOUT_MS = 1000 * 60 * 20
```

File: software/contrib/quantizer.py
Comment (awonak) [reply software/contrib/quantizer.py:471]: Ah, sounds like it's a "feature" then. Makes sense to me.

File: software/contrib/quantizer.py
Comment (awonak) [reply software/contrib/quantizer.py]: Ha, async overlapping comments :)

Double check the order of time values in your call to `ticks_diff`. I believe you might need to reverse them according to the docs examples for "Scheduling Events" - https://docs.micropython.org/en/latest/library/time.html#time.ticks_diff

Comment (chrisib) [reply software/contrib/quantizer.py]: Just double-checked with a 10s timeout. The order in the latest commit does work as intended; after 10s of being idle the screen shuts off, and wakes back up on a button press.  I'm testing again with the full 20-minute timeout, but I don't expect it to behave differently.

I based it on the timeout example:
```
# Wait for GPIO pin to be asserted, but at most 500us
start = time.ticks_us()
while pin.value() == 0:
    if time.ticks_diff(time.ticks_us(), start) > 500:
        raise TimeoutError
```
which is equivalent to shutting the screen off after the timeout elapses:
```
now = time.ticks_us()
if time.ticks_diff(now, self.last_interaction_time) > SCREENSAVER_TIMEOUT_US:
    self.active_screen = self.screensaver
```
where the current time is the same and `start` is equivalent to `last_interaction_time`

Comment (chrisib) [reply software/contrib/quantizer.py]: There was a bug with the 20 minute timeout.  Looks like there may have been an integer overflow when dealing with intervals that big.  Switching to ms seems to have fixed it though.

Comment (awonak) [reply software/contrib/quantizer.py]: Thanks for confirming and testing. LGTM.

**Review by awonak — APPROVED** [review]
LGTM

**Review by mjaskula — CHANGES_REQUESTED** [review]
I've been wanting a EuroPi quantizer for a while now, so this PR is very welcome. Thank you for contributing it.

If you have't read it yet, see the contributing guide so that you know what to expect from reviewers.  https://github.com/Allen-Synthesis/EuroPi/blob/main/contributing.md

I've added some comments with some suggestions, questions, and discussion points. Most are optional, take what you like and ignore the rest. There is one small documentation change that I think should be made before this is merged.

Let us know when you are finished considering these suggestions and we can get this merged. Thanks again.

File: software/contrib/quantizer.py
Comment (chrisib) [reply software/contrib/quantizer.py:451]: On the outside chance someone needs the Quantizer class as a utility inside some other script, I didn't want to clobber their button/trigger handlers.  Separating the I/O connections means you can technically instantiate the Quantizer class to access the `.quantize(...)` method, without needing to worry about this class interfering with whatever other event-handlers you're using.

File: software/contrib/quantizer.md
Comment (chrisib) [reply software/contrib/quantizer.md:1]: Done

File: software/contrib/quantizer.py
Comment (chrisib) [reply software/contrib/quantizer.py]: Converted class/function comment blocks with docstrings.

File: software/contrib/quantizer.py
Comment (chrisib) [reply software/contrib/quantizer.py]: Good call.

File: software/contrib/quantizer.py
Comment (chrisib) [reply software/contrib/quantizer.py]: That's a leftover from an older revision. Good catch.

File: software/contrib/quantizer.py
Comment (chrisib) [reply software/contrib/quantizer.py]: Renamed & streamlined all of these functions.

File: software/contrib/quantizer.py
Comment (chrisib) [reply software/contrib/quantizer.py:17]: The three options I see are:

- we merge the Quantizer as-is, and it can continue using the existing code, oblivious to the changes in #220
- we merge the Quantizer now, and once #220 is merged I/someone can update it to use the new config system
- we hold off merging in the Quantizer until after #220 is merged, I rebase, and update this merge request

File: software/contrib/quantizer.py
Comment (chrisib) [reply software/contrib/quantizer.py]: Done.

File: software/contrib/quantizer.py
Comment (mjaskula) [reply software/contrib/quantizer.py:17]: Definitely option 1 or 2. I just wanted to make you aware, in case it was something you were interested in.

Comment (civerachb-cpr) [reply software/contrib/quantizer.py:17]: I'll keep an eye on what changes and see about submitting some updates.  The 3 new scripts I've submitted have some code copied between them I'd like to optimize out into some sort of shared library eventually, too.

Comment (chrisib) [reply software/contrib/quantizer.py:17]: I'll keep an eye on what changes and see about submitting some updates. The 3 new scripts I've submitted have some code copied between them I'd like to optimize out into some sort of shared library eventually, too.

File: software/contrib/quantizer.py
Comment (mjaskula) [reply software/contrib/quantizer.py:434]: I don't think that is an outside chance at all. In fact I'm pretty sure that I will be using this class in some future work. 

If you truly want to make a reusable quantizer class, I would suggest that it not be the class that extends `EuroPiScript`. Instead, I would have a regular class that provides quantizer functionality, and then a second class that ties it to the EuroPi hardware and `EuroPiScript` hierarchy. I tried to do a similar thing with the `TuringMachine` class. Though admittedly, the class that extends EuroPiScript is pretty convoluted.

If you do want to do this, I would suggest that it happens in a follow up PR, as this one is nearly ready to be merged.

**Review by mjaskula — APPROVED** [review]
Thanks for making those changes. I think that this is ready to go out to the masses.

File: software/contrib/quantizer.py
Comment (mjaskula) [inline software/contrib/quantizer.py:530]: [discussion] I'm seeing some strange behavior, that I'm having trouble explaining. 

Here's my setup: I have the quantizer patched up in such a way that that I'm feeding it a constant voltage to ain that I can control via a knob on another module. I'm sending cv1 to the pitch of an oscillator and am listing to the output. I then tune the input voltage to a place where the quantizer has trouble choosing between two notes. It's easier to do this if you turn off a bunch of notes in the scale and put the voltage in the large empty space you created.

The fact that the note jumps around is not the strange part. It is known that the voltages read by ain on this hardware can be noisy, and the code does nothing to mitigate it. We can iterate on this later to attempt to tame it.

The strangeness is that the note jumps around at a different rate depending on which screen I'm on. I've been able to 'tune' it such that the note is stable when viewing the piano, but erratic when viewing the menu, and vice versa.

I've looked through the code and don't see anything in the menu that would directly affect the quantization. Since you are considerably more familiar with this code, I thought that I would point this out to you in case you had any ideas. 

My only idea is that the drawing of the piano takes more time than the menu, causing a slightly different rate of quantization depending on what screen is being viewed. And this rate somehow interacts with something else on the pico, such as the rate of whatever is causing the ain noise.

I don't think that this necessarily needs to be fixed before merge, but it is strange enough that I thought that it might point towards a subtle bug that I just can't see yet.

File: software/contrib/quantizer.py
Comment (mjaskula) [inline software/contrib/quantizer.py:523]: [question] these two lines cause k1 to be read twice, when only one value will be used on any particular trip through the loop (depending on which screen is being viewed). Why do the reads here instead of where they are being used, as you do with knob 2? I feel like i am missing something.

File: software/contrib/quantizer.py
Comment (mjaskula) [inline software/contrib/quantizer.py:510]: [discussion] I feel like this comment is misleading. I don't think that we know enough about the rate of execution of the rest of the loop to make a claim about how fast we are quantizing.

File: software/contrib/quantizer.py
Comment (mjaskula) [reply software/contrib/quantizer.py:512]: > It's there mostly to ensure that the trigger signal generated by the continuous quantization mode is on for at least 10ms.

That sounds important, let's add a comment to that effect, so that we don't forget it when making changes in the future.

Would it make sense to move this sleep closer to the continuous quantization code? Maybe event into the `if` that turns cv6 on? or would this break something else?

**Review by mjaskula — CHANGES_REQUESTED** [review]
@chrisib Thanks for pulling out the Scale class so that we can more easily use it elsewhere. I'm looking forward to it.

I did find one bug in the new code that will need to be fixed before merge. I've also opened up a few discussion points on topics that I noticed in my re-review. I look forward to hearing your opinions.

**Comment by awonak** [comment]
I did some testing with my Mordax Data voltage monitor and the tuner. I noticed some unexpected behavior that I wanted to discuss.

I have set my EuroPi Quantizer to a C Major scale (C, E, G).

When I feed 0v into the EuroPi Quantizer Ain and tune my oscillator to C1 and then increase the Ain voltage from 0v to 1v the tuner shows the oscillator moves to F#1 (6 semitones?). I would expect it to tune to C2.

However, if I feed 1v into the EuroPi Quantizer Ain and tune my oscillator to C1 and then increase the Ain voltage from 1v to 2v, then the tuner shows the oscillator moves near C2 as expected. I can continue to move voltage between 1v and 5v and the tuner shows roughly in tune C, E, G notes as expected.

This suggests there is some bug in the quantization around 0v, perhaps one of those off-by-one errors? @chrisib do you happen to have a Data or other decent tools for measuring voltage and tune to troubleshoot this?

File: software/contrib/quantizer.py
Comment (awonak) [reply software/contrib/quantizer.py:533]: One possible solution here is to reduce the frequency of draw calls by introducing a "dirty" flag to indicate if the UI has changed, and only draw the screen if "dirty" is `True`. You can see as similar implementation in the Polyrhythmic Seq script where I prevent unnecessary saving if state has not changed or minimum duration has not yet passed: https://github.com/Allen-Synthesis/EuroPi/blob/759478558332e47552f34cd2b4e5cd5d16df1fb7/software/contrib/polyrhythmic_sequencer.py#L236

**Comment by chrisib** [comment]
Regarding [this question](https://github.com/Allen-Synthesis/EuroPi/pull/222#discussion_r1108607889), it can (and how has) been moved and an explanatory comment inserted.

**Comment by chrisib** [comment]
> I did some testing with my Mordax Data voltage monitor and tuner. I noticed some unexpected behavior that I wanted to discuss.
> 
> I have set my EuroPi Quantizer to a C Major scale (C, E, G).
> 
> When I feed 0v into the EuroPi Quantizer Ain and tune my oscillator to C1 and then increase the Ain voltage from 0v to 1v the tuner shows the oscillator moves from C1 to F#1 (6 semitones?). I would expect it to tune to C2.
> 
> However, if I feed 1v into the EuroPi Quantizer Ain and tune my oscillator to C1 and then increase the Ain voltage from 1v to 2v, then the tuner shows the oscillator moves from C1 to C2 as expected. I can continue to move voltage between 1v and 5v and the tuner shows roughly in tune C, E, G notes as expected. If the voltage drops from 1v to 0v the tuner reads F#1, not C0 as expected.
> 
> This suggests there is some bug in the quantization around 0v, perhaps one of those off-by-one errors? @chrisib do you happen to have a Data or other decent tools for measuring voltage and tune to troubleshoot this?

I suppose the obvious first question: have you gone through the precise calibration procedure for your EuroPi?  I unfortunately do not have a high-precision 10V power supply and a decent scope to debug things at the per-volt level.  I expected there may be some bugs with the actual voltage stuff, and was hoping someone with access to better equipment than me could help with some of the testing.

I just tried replicating your conditions.  I've got a channel of Quadratt plugged into `ain` via Zeroscope so I can get a coarse sense of what's happening.  I have my oscillator tuned to C2.  EuroPi is configured to quantize to C/E/G, just like yours.

Watching the keyboard display on EuroPi, I turn the knob on Quadratt to increase the voltage to `ain` until I see EuroPi up a full octave.  The tuner shows the oscillator is playing B3, about one semitone flat from where it should be.  Going up another octave on the keyboard display I get to A4# on the tuner.  One more octave on the keyboard gives me A5 on the tuner.  So it's consistently losing ~1 semitone per octave, but I'll chalk that up to my EuroPi not being well-calibrated.

I unfortunately do not have a good way to determine what the actual output voltage from Quadratt is beyond squinting at a low-res line on Zeroscope.  The line looks to be just shy of 1V when EuroPi's keyboard says I've gone up from C to C, which is consistent with everything else I've replicated.  Going up another octave the line is somewhere in the vicinity of 2V, but the resolution beyond this point is so imprecise that it's meaningless.

I'll see if I can get some better-quality equipment to debug this more.  But in the meantime, if you could make sure you've (re-)calibrated your EuroPi and repeat your experiment I'd appreciate it.  And if anyone else wants to try reproducing these results, more data points would always be good.

File: software/contrib/quantizer.py
Comment (chrisib) [reply software/contrib/quantizer.py]: That was a holdover from an earlier version where more work was being done inside the main loop. Things got re-organized, and that part just got ignored because it was working.

I've relocated reading k1 now

File: software/contrib/quantizer.py
Comment (chrisib) [reply software/contrib/quantizer.py:507]: Yeah, the noise-mitigation is an issue. The only thing I've really done to address it was to increase the samples when reading `ain` in an attempt to smooth it slightly. But if your voltage is on the boundary between rounding up vs rounding down then you will get some jitter with the current implementation.

I wouldn't be surprised if the keyboard takes significantly longer to draw; it's a bitmap layer with text super-imposed over it with a couple of loops, as opposed to _just_ some text.  I suspect the bit-blitting to get the keyboard background layer uses most of the difference in CPU cycles.

De-noising is definitely something that's in the queue for the next iteration. Or if someone with more experience doing noise-reduction on a low-power system wants to tackle it I'm happy to accept MRs back to my source branch too.

**Comment by awonak** [comment]
> I suppose the obvious first question: have you gone through the precise calibration procedure for your EuroPi?

With egg on my face I admit that EuroPi was uncalibrated and that was the problem! After performing the high precision calibration process, the quantizer behaves exactly as expected! 0v tuned to C2 stepped through all notes in the quantized C Maj scale as expected!

**Comment by chrisib** [comment]
> With egg on my face I admit that my EuroPi was uncalibrated and that was the problem! After performing the high precision calibration process, the quantizer behaves exactly as expected! 0v tuned to C2 stepped through all notes in the quantized C Maj scale as expected!

Haha, I'm glad it was the easy solution and not some weird systemic bug with the software!

> https://photos.app.goo.gl/xvk54bXgxdeYgqxp8
That's a great little video. I'm glad it's working correctly for you now :)

File: software/contrib/quantizer.py
Comment (mjaskula) [inline software/contrib/quantizer.py:303]: Great idea here.

**Review by mjaskula — APPROVED** [review]
Sorry for the delay in reviewing these changes, I was on vacation last week.

These changes look great, I think it's ready to merge.

**Pull request merged**
